cmd/devtools/customlogs/customlogs/infer.go

```
@@ -0,0 +1,466 @@
+	File     *string
+	NoVerify *bool
+}
+
+var inferJsoniter = jsoniter.Config{
+	UseNumber: true,
+}.Froze()
+
+// Infers a schema given a sample of logs
+func Infer(logger *zap.SugaredLogger, opts *InferOpts) {
+	if *opts.File == "" {
+		flag.Usage()
+		logger.Fatal("no sample file provided")
+	}
+
+	schema, err := inferFromFile(logger, *opts.File)
+	if err != nil {
+		logger.Fatal("failed to generate schema", zap.Error(err))
+	}
+
+	if !*opts.NoVerify {
+		// In order to validate that the schema generated is correct,
+		// run the parser against the logs, fail in case of error
+		if err = validateSchema(schema, *opts.File); err != nil {
+			logger.Fatal("failed while testing schema with file", zap.Error(err))
+		}
+	}
+
+	marshalled, err := yaml.Marshal(schema)
+	if err != nil {
+		logger.Fatal("failed to marshal schema", zap.Error(err))
+	}
+	fmt.Println(string(marshalled))
+}
+
+func inferFromFile(logger *zap.SugaredLogger, file string) (logschema.Schema, error) {
+	schema := logschema.Schema{
+		Version: 0,
+	}
+	fd, err := os.Open(file)
+	if err != nil {
+		return schema, errors.Wrap(err, "failed to read input file")
+	}
+	defer fd.Close()
+
+	scanner := bufio.NewScanner(fd)
+	lineNum := 0
+	for scanner.Scan() {
+		lineNum++
+		line := bytes.TrimSpace(scanner.Bytes())
+		if len(line) == 0 {
+			continue
+		}
+
+		var data map[string]interface{}
+		if err = inferJsoniter.Unmarshal(line, &data); err != nil {
+			return schema, errors.Wrapf(err, "failed to parse line [%d] as JSON", lineNum)
+		}
+		// inferring the log schema of that single line
+		lineFields := inferFields(data)
+		// merging the schema of this line with the schema that we have generated from all lines until now
+		schema.Fields, err = mergeFields(schema.Fields, lineFields)
+		if err != nil {
+			return schema, errors.Wrapf(err, "failed while inferring schema from line [%d]", lineNum)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		logger.Fatal("failed to read input file", zap.Error(err))
```

Since this function has an error return argument, why not return the error and let it be handled in one place higher in the stack?

---

Ah yes. I missed that one. Removed.